gtp2ogs.js

```diff
@@ -1017,10 +1021,36 @@ class Game {
         if (this.state.phase != 'play')
             return;
 
-	let sendMove = (moves) => {  this.uploadMove(moves[0]);  };
-	let sendPass = ()      => {  this.uploadMove({'x': -1});  };
+	let sendPass = () => {  this.uploadMove({'x': -1});  };
+	let doing_handicap = (this.state.free_handicap_placement && this.state.handicap > 1 &&
+			      this.state.moves.length < this.state.handicap);
 	
-	this.getBotMoves("genmove " + this.my_color, sendMove, sendPass);
+	if (!doing_handicap) {  // Regular genmove ...
+	    let sendMove = (moves) => {  this.uploadMove(moves[0]);  };
+	    this.getBotMoves("genmove " + this.my_color, sendMove, sendPass);
+	    return;
+	}
+	
+	// Already have handicap stones ? Return next one.
+	if (this.handicap_moves.length) {
+	    this.uploadMove(this.handicap_moves.shift());
+	    return;
+	}
+
+	// Get handicap stones from bot and return first one.
+	let storeMoves = (moves) => {
+	    if (moves.length != this.state.handicap) {  // Sanity check
+		this.log("place_free_handicap returned wrong number of handicap stones, resigning.");
+		if (this.bot) this.bot.kill();
+		this.bot = null;
+		this.uploadMove({'resign': true});
+		return;
+	    }
+	    this.handicap_moves = moves;
```

Comment:
The bot I'm working on used to generate a PASS as a handicap move from time to time - if given way too many handicap moves to play. In that case, `gtp2ogs` would just hang as the server will ignore the pass as a handicap.

I think with this commit, that also happens because we upload the pass and the server will (correctly) reject it.

Arguably, bots should implement the spec correctly, but it would be helpful to bot authors if `gtp2ogs` would check that there are not `pass` moves and complain and resign the game if there are any. Large handicaps come up rarely so it is relatively easier to miss that case during development.

Reply:
I'd say same, don't use gtp2ogs to fix bugs in your bot. Fix your bot instead ! Or do some workarounds but keep it a secret =)

Reply:
as said in other comment :+1: 

phoenixgo supports free handicap as is currently on devel branch, as black and as white without any issue (to give or take handi)

so phoenixgo does not need this issue, it works fine already
also, phoenix go is unlikely to take handicap from anyone on ogs (also its bad at handi to begin with)

the only thing i pointed is the behaviour of human waiting long or leaving games waiting for bot to play while they didnt put all handicap stones, as explained in #80 

you are right that we shouldnt hack gtp2ogs, custom branches can achieve that

thanks for your effort everyone :+1:

Reply:
Well, you check above that the correct number of handicap stones is returned by the bot, to avoid getting into the state where you send a partial handicap to the server and then get stuck forever (if I understand the code right).

Sending one or more `pass` moves as part of handicap is the same as sending a shorter handicap and `gtp2ogs` will get stuck. The user will wait for the bot to move and nothing will happen. The game will time out and the player will be annoyed.

So we should just validate the input from the bot here and fail fast if the input was incorrect. The spec says:

https://www.lysator.liu.se/~gunnar/gtp/gtp2-spec-draft2/gtp2-spec.html#SECTION00073200000000000000

```
Vertices must not be repeated or include ``pass''.
```

Reply:
Yes, we should do this. Sorry for misunderstanding your previous comment, this is exposing bugs in bots, not hiding them.